commons/src/main/java/io/aiven/kafka/tieredstorage/commons/UniversalRemoteStorageManager.java

```diff
@@ -16,54 +16,263 @@
+        try {
+            return Cipher.getInstance("AES/GCM/NoPadding", "BC");
+        } catch (final NoSuchAlgorithmException | NoSuchProviderException | NoSuchPaddingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected static Cipher decryptionCipherSupplier(final byte[] encryptedChunk,
+                                                     final SegmentEncryptionMetadata encryptionMetadata) {
+        try {
+            final Cipher encryptCipher = getCipher();
+            encryptCipher.init(Cipher.DECRYPT_MODE, encryptionMetadata.secretKey(),
+                new IvParameterSpec(encryptedChunk, 0, encryptionMetadata.ivSize()),
+                SecureRandom.getInstanceStrong());
+            encryptCipher.updateAAD(encryptionMetadata.aad());
+            return encryptCipher;
+        } catch (final NoSuchAlgorithmException | InvalidKeyException | InvalidAlgorithmParameterException e) {
+            throw new RuntimeException(e);
+        }
     }
 
     @Override
     public InputStream fetchLogSegment(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                        final int startPosition) throws RemoteStorageException {
-        throw new UnsupportedOperationException();
+        return this.fetchLogSegment(remoteLogSegmentMetadata, startPosition,
+            remoteLogSegmentMetadata.segmentSizeInBytes());
     }
 
     @Override
     public InputStream fetchLogSegment(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                        final int startPosition,
                                        final int endPosition) throws RemoteStorageException {
-        throw new UnsupportedOperationException();
+        try {
+            final InputStream manifest = objectStorageFactory.fileFetcher()
+                .fetch(objectKey.key(remoteLogSegmentMetadata, ObjectKey.Suffix.MANIFEST));
+
+            final SegmentManifestV1 segmentManifestV1 = mapper.readValue(manifest, SegmentManifestV1.class);
+            final FetchChunkEnumeration fetchChunkEnumeration =
+                new FetchChunkEnumeration(chunkManager, remoteLogSegmentMetadata, segmentManifestV1, startPosition,
+                    endPosition);
+            return new SequenceInputStream(fetchChunkEnumeration);
+        } catch (final IOException e) {
+            throw new RemoteStorageException(e);
+        }
     }
 
     @Override
     public InputStream fetchIndex(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                   final IndexType indexType) throws RemoteStorageException {
-        throw new UnsupportedOperationException();
+        try {
+            return objectStorageFactory.fileFetcher()
+                .fetch(objectKey.key(remoteLogSegmentMetadata, ObjectKey.Suffix.fromIndexType(indexType)));
+        } catch (final IOException e) {
+            throw new RemoteStorageException(e);
+        }
+
```

Comment:
I tried running this against the current TS broker implementation and it fails as it tries to fetch `IndexType.TRANSACTION`, which may or may not exist. It seems, it expects `null` if it doesn't exist:
https://github.com/aiven/kafka/blob/70c4fb03da91ba8c9b74af2200696e7db5347c6c/core/src/main/scala/kafka/log/remote/RemoteIndexCache.scala#L185

I think we should return `null` (if there's indeed no transaction index) and also leave a TODO comment here (or a ticket) that this needs to be aligned with the future upstream implementation.

Reply:
The doc is not that clear:
```RemoteResourceNotFoundException – when there are no resources associated with the given remoteLogSegmentMetadata```
I read it that way that "no resources" = "the segment doesn't exist at all". This may matter because in the upstream implementation (now or at some point) some decisions that we don't want may be made based on this exception

Reply:
We hsould decide something on this. Looks like @jeqo is right here but on the other hand it is unclear how to distinguish lack of transaction index when it is not configured vs missing transaction issues in case of any issues.

Reply:
Looking for clarification here: https://issues.apache.org/jira/browse/KAFKA-7739?focusedCommentId=17720936&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17720936

Reply:
Return `null` + leave a TODO for now?

Reply:
Interestingly, upstream doesn't handle this exception as in the linkedin fork: https://github.com/apache/kafka/blob/322ac86ba282f35373382854cc9e790e4b7fb5fc/core/src/main/scala/kafka/log/remote/RemoteIndexCache.scala#L197-L211
and if returning null, it will break as exception is not handled AFAICS.

Reply:
In the LInkedIn fork AFAIU `remoteStorageManager.fetchIndex` is wrapped into `Option` for that specific case(e.g. null was returned)

Reply:
Seems if we are sticking to LinkedIn version for now then null should probably be fine. So proceeding with @ivanyu's suggestion